### PR TITLE
*: support show auto increment ID

### LINKED ISSUE #4678: show current auto_increment ID
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
```
create table t(id int primary key auto_increment);
insert into t values(1);
show create table t;
```

2. What did you expect to see?
```
+-------+-----------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                            |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=latin1 |

```

3. What did you see instead?
```
+-------+---------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                          |
+-------+---------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin |
+-------+---------------------------------------------------------------------------------------------------------------------------------------+

```

4. What version of TiDB are you using (`tidb-server -V`)?

### PR DESCRIPTION
**Pull request by zimulala**
Fix #4678 
And fix the `show create table show_auto_increment` result of `create table show_auto_increment (id int) auto_increment=4`.  #3747

### CODE REVIEW

File: executor/show_test.go [hunk 1]
```diff
@@ -175,6 +175,31 @@ func (s *testSuite) TestShow(c *C) {
 			"  `id` int(11) DEFAULT NULL\n"+
 			") ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin AUTO_INCREMENT=4",
 	))
+	// for issue https://github.com/pingcap/tidb/issues/4678
+	tk.MustExec("insert into show_auto_increment values(10)")
+	tk.MustQuery(`show create table show_auto_increment`).Check(testutil.RowsWithSep("|",
```
Comment (shenli) [inline executor/show_test.go]: If the table doesn't have a column with auto_increment attribute, MySQL will ignore the table option.
```sql
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 89
Server version: 5.7.20-0ubuntu0.16.04.1 (Ubuntu)

Copyright (c) 2000, 2017, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> create table show_auto_increment (id int) auto_increment=4;
Query OK, 0 rows affected (0.02 sec)

mysql> show create table show_auto_increment;
+---------------------+---------------------------------------------------------------------------------------------------------+
| Table               | Create Table                                                                                            |
+---------------------+---------------------------------------------------------------------------------------------------------+
| show_auto_increment | CREATE TABLE `show_auto_increment` (
  `id` int(11) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=latin1 |
+---------------------+---------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> 
```

Comment (zimulala) [reply executor/show_test.go]: Yep, but it's the old logical in TiDB. It handles in PR 3747.
I will fix it.

File: executor/show_test.go [hunk 2]
```diff
@@ -168,13 +168,40 @@ func (s *testSuite) TestShow(c *C) {
...
 			") ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin AUTO_INCREMENT=4",
 	))
+	// for issue https://github.com/pingcap/tidb/issues/4678
+	tk.MustExec("insert into show_auto_increment values(20)")
+	tk.MustQuery(`show create table show_auto_increment`).Check(testutil.RowsWithSep("|",
+		""+
+			"show_auto_increment CREATE TABLE `show_auto_increment` (\n"+
+			"  `id` int(11) NOT NULL AUTO_INCREMENT,\n"+
+			"  PRIMARY KEY (`id`)\n"+
+			") ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin AUTO_INCREMENT=5021",
```
Comment (winkyao) [inline executor/show_test.go]: Plz add some comment here, about why this is 5011, not 11, autoid step is 5000 in `meta/autoid/autoid.go:30`

File: executor/show_test.go [hunk 3]
```diff
@@ -168,13 +168,40 @@ func (s *testSuite) TestShow(c *C) {
...
+			"  PRIMARY KEY (`id`)\n"+
+			") ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin",
+	))
+	tk.MustExec("insert into show_auto_increment values(10)")
+	tk.MustQuery(`show create table show_auto_increment`).Check(testutil.RowsWithSep("|",
+		""+
+			"show_auto_increment CREATE TABLE `show_auto_increment` (\n"+
+			"  `id` int(11) NOT NULL AUTO_INCREMENT,\n"+
+			"  PRIMARY KEY (`id`)\n"+
+			") ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin AUTO_INCREMENT=5011",
```
Comment (winkyao) [inline executor/show_test.go]: ditto

File: executor/show_test.go (hunk 2 above)
Comment (winkyao) [reply executor/show_test.go]: or, we can exported the step in autoid, and use it in here.

File: util/kvencoder/allocator.go [hunk 4]
```diff
@@ -50,3 +50,7 @@ func (alloc *allocator) Base() int64 {
 func (alloc *allocator) End() int64 {
 	return alloc.Base() + step
 }
+
+func (alloc *allocator) NextGlobalAutoID(tableID int64) (int64, error) {
+	return alloc.Base() + step + 1, nil
```
Comment (winkyao) [inline util/kvencoder/allocator.go]: return alloc.End() + 1

File: meta/autoid/autoid.go [hunk 5]
```diff
@@ -76,6 +78,18 @@ func (alloc *allocator) End() int64 {
 	return alloc.end
 }
 
+// NextGlobalAutoID implements autoid.Allocator NextGlobalAutoID interface.
+func (alloc *allocator) NextGlobalAutoID(tableID int64) (int64, error) {
+	var autoID int64
+	err := kv.RunInNewTxn(alloc.store, true, func(txn kv.Transaction) error {
+		var err1 error
+		m := meta.NewMeta(txn)
+		autoID, err1 = m.GetAutoTableID(alloc.dbID, tableID)
```
Comment (winkyao) [inline meta/autoid/autoid.go]: `autoID, err1 := m.GetAutoTableID(alloc.dbID, tableID)`, remove line 85

Comment (zimulala) [reply meta/autoid/autoid.go]: I need to return autoID at the end of `NextGlobalAutoID `.

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
@winkyao PTAL

**Comment by shenli** [comment]
LGTM

**Pull request merged**